src/ray/raylet/node_manager.cc

```diff
@@ -2046,6 +2029,16 @@ void NodeManager::ProcessSetResourceRequest(
 bool NodeManager::PrepareBundle(
     std::unordered_map<ClientID, SchedulingResources> &resource_map,
     const BundleSpecification &bundle_spec) {
+  // We will first delete the existing bundle to ensure idempotent.
+  // The reason why we do this is: after GCS restarts, placement group can be rescheduled
+  // directly without rolling back the operations performed before the restart.
+  const auto &bundle_id = bundle_spec.BundleId();
+  if (bundle_state_map_.contains(bundle_id)) {
```

Comment:
Actually, this can cause problems if the bundle was already committed because `CancelResourceServe` also kills workers that take up a bundle resources. We should probably also check if bundle state is COMMIT or not.

Reply:
Yeah, you are right. But I have some questions:
1. Why `HandleCancelResourceReserve` kill all workers that are currently associated with the placement group? I think we should  just kill the worker corresponding to the bundle.
2.We will kill worker process, if there is an actor in the worker process that does not belong to this PG, it will not be restart  because it is intentional_disconnect is true.

Reply:
1) That's a good point. It must be a mistake (and I think in terms of consequences, they are the same, and that's why we didn't notice :( ). We should kill only workers associated with the bundle. 
2) I don't really understand what this means. We are killing workers that are only associated with this PG (and nothing else).

Reply:
> 1. That's a good point. It must be a mistake (and I think in terms of consequences, they are the same, and that's why we didn't notice :( ). We should kill only workers associated with the bundle.
> 2. I don't really understand what this means. We are killing workers that are only associated with this PG (and nothing else).

Do you mean that the worker process will be monopolized by PG? If it is not exclusive, the kill worker process will cause the actors above it not to be restarted.

Reply:
I think it makes sense that pg monopolizes actors because if the pg is removed, the actor will never be restartable because resources won't be available forever (if I understood you correctly).

Reply:
I want to submit a PR separately, kill only workers associated with the bundle. And i will move the logic of kill worker(https://github.com/ray-project/ray/pull/10998/files#diff-50e88810cd198c83bbce3317b1bc9924L1855) to `ReturnBundleResources`.

Reply:
Before that, I'll add a node_manager UT first.

Reply:
@rkooo567 Can this PR be merged first?

Reply:
I think we shouldn't kill workers when the bundle was already committed when prepare requests are received. We should just return PrepareRequest.

Reply:
Because if the bundle state is already committed, it means that prepare requests are just stale.

Reply:
Oh, i got it, i will fix it. thanks.